src/python/pants/backend/codegen/protobuf/scala/rules.py

```diff
@@ -0,0 +1,163 @@
+        Get(DownloadedExternalTool, ExternalToolRequest, protoc.get_request(Platform.current)),
+        Get(
+            MaterializedClasspath,
+            MaterializedClasspathRequest(
+                prefix="__toolcp",
+                lockfiles=(scalapbc.resolved_lockfile(),),
+            ),
+        ),
+        Get(Digest, CreateDigest([Directory(output_dir)])),
+        Get(TransitiveTargets, TransitiveTargetsRequest([request.protocol_target.address])),
+    )
+
+    # NB: By stripping the source roots, we avoid having to set the value `--proto_path`
+    # for Protobuf imports to be discoverable.
+    all_sources_stripped, target_sources_stripped, protoc_gen_scala_digest = await MultiGet(
+        Get(
+            StrippedSourceFiles,
+            SourceFilesRequest(
+                tgt[ProtobufSourceField]
+                for tgt in transitive_targets.closure
+                if tgt.has_field(ProtobufSourceField)
+            ),
+        ),
+        Get(
+            StrippedSourceFiles, SourceFilesRequest([request.protocol_target[ProtobufSourceField]])
+        ),
+        Get(
+            Digest,
+            CreateDigest(
+                [
+                    FileContent(
+                        "protoc-gen-scala",
+                        textwrap.dedent(
+                            f"""\
+                                       #!{bash.path}
+                                       exec {' '.join(jdk_setup.args(bash, tool_classpath.classpath_entries())[1:])} scalapb.ScalaPbCodeGenerator "$@"
+                                       """
+                        ).encode(),
+                        is_executable=True,
+                    )
+                ]
+            ),
+        ),
+    )
+
+    unmerged_digests = [
+        all_sources_stripped.snapshot.digest,
+        downloaded_protoc_binary.digest,
+        tool_classpath.content.digest,
+        protoc_gen_scala_digest,
+        empty_output_dir,
+        jdk_setup.digest,
+    ]
+
+    input_digest = await Get(Digest, MergeDigests(unmerged_digests))
+
+    args = [
+        downloaded_protoc_binary.exe,
+        "--plugin=protoc-gen-scala=./protoc-gen-scala",
+        "--scala_out",
+        output_dir,
+        *target_sources_stripped.snapshot.files,
+    ]
+
+    result = await Get(
+        ProcessResult,
+        Process(
+            args,
+            input_digest=input_digest,
+            description=f"Generating Scala sources from {request.protocol_target.address}.",
+            level=LogLevel.DEBUG,
+            output_directories=(output_dir,),
+            env=jdk_setup.env,
+            append_only_caches=jdk_setup.append_only_caches,
+        ),
+    )
+
+    normalized_digest, source_root = await MultiGet(
+        Get(Digest, RemovePrefix(result.output_digest, output_dir)),
+        Get(SourceRoot, SourceRootRequest, SourceRootRequest.for_target(request.protocol_target)),
```

Comment:
Note: I left out the support for specifying an alternate source root. It was not clear whether that was something only Python needed.

cc @Eric-Arellano